### ISSUE TITLE
Implement SQLITE lockdown state

### ISSUE THREAD

**Issue by evilaliv3** [globaleaks/GlobaLeaks#3206]
The [Defense Against The Dark Arts](https://www.sqlite.org/security.html) guide of SQLITE includes various suggestions about some features that could be used to implement a "lockdown" state of SQLALCHEMY/SQLITE functionalities useful to limit the exploitation possibilities in in case of SQL Injection.

Among the suggestion is the use of sqlite3 ``set_authorizer`` function and of ``trusted_schema=OFF`` pragma that makes it possible to restrict the type of queries.

It would be probably interesting to check this aspect more precisely to evaluate the possible added value in our context of use.

I consider we could benefit of those by implementing a lockdown state to be enabled after preparing the database:
```python
    def authorizer_callback(action, table, column, sql_location, ignore):
        if action in [SQLITE_DELETE,
                      SQLITE_INSERT,
                      SQLITE_READ,
                      SQLITE_SELECT,
                      SQLITE_TRANSACTION,
                      SQLITE_UPDATE] or \
           (action == SQLITE_FUNCTION and column in ['count',
                                                     'lower',
                                                     'min',
                                                     'max']):
            return sqlite3.SQLITE_OK
        else:
            return sqlite3.SQLITE_DENY
```

This would for example limit the type of SQL Queries executed ensuring that SQLITE will always ask for confirmation and possibly prevent exploitation in case of possible new zero-day like [CVE-2019-7164](https://nvd.nist.gov/vuln/detail/CVE-2019-7164) and [CVE-2019-7548](https://nvd.nist.gov/vuln/detail/CVE-2019-7548) and usage of advanced technichs like what prevented by Omer Gull at 36c3 in 2019: https://media.ccc.de/v/36c3-10701-select_code_execution_from_using_sqlite

The example above for example escludes the possibility to execute PRAGMAs, full TABLE DELETION, and creation of VIRTUAL TABLES and the use of all the auxiliary SQLITE function like fts3_tokenizer that are known pre-requirements for a set of known advanced exploitation techniques.

**Comment by evilaliv3** [comment]
@zzzeek: Do you have any suggestion based on your experience? Do you know projects that have implemented similar approaches?

@conorsch: have you ever evaluated a similar approach in the context of @securedrop?

**Comment by zzzeek** [comment]
while it's good if your database is protected against DROP statements, SQL injection attacks don't need DROP to blow away the database, all it needs is DELETE or UPDATE.   Similarly, "unusually large inputs", it's of course good to limit how big a SQL statement can be, but if I had control of SQL statements I could just run the same INSERT statement thousands of times until the machine runs out of disk space.   

that said of course you can set all these things up provided the Python sqlite3 interface gives you access to them, such as sqlite3_limit(), etc.    set them up in a [connect](https://docs.sqlalchemy.org/en/14/core/events.html?highlight=poolevents%20connect#sqlalchemy.events.PoolEvents.connect) handler.

I've not seen this page on sqlite's website before and I haven't seen anyone trying to do this, but it should be straightforward.

**Comment by evilaliv3** [comment]
Thank you so much for your feedback @zzzeek.

I agree on your points about DROP vs UPDATE.

Our main interest in these approach is to limit escalation from sqlinjections to code execution and is just for this reason that in my example i considered that we could authorize only thefunctionalities that we effectively use within our own codebase.

**Comment by evilaliv3** [comment]
@zenmonkeykstop: in commit https://github.com/globaleaks/GlobaLeaks/commit/5fa213d538f168cdbbf7b29ad634d79696fbf585 you find a generic implementation for the additional protection layer we were discussing.

I invite you to give it a try as i consider this could work like a charm in securedrop with minor changes.

**Comment by evilaliv3** [comment]
Implemented!

This additional protection layer will be included starting from upcoming release 4.9.0
